File: src/api/Services/BaseService.cs
using Example.Api.Dtos.Responses;
using Example.Api.Enums;

namespace Example.Api.Services;

/// <summary>
/// Base class for services.
/// </summary>
public abstract class BaseService
{
    /// <summary>
    /// Creates a successful API result.
    /// </summary>
    /// <param name="message">Optional success message.</param>
    /// <returns>A successful ApiResult instance.</returns>
    protected ApiResult SuccessResult(string? message = default)
    {
        return new ApiResult
        {
            Success = true,
            Code = ApiCode.Success,
            Message = message,
        };
    }

    /// <summary>
    /// Creates a successful API result with data.
    /// </summary>
    /// <param name="data">The data to include in the result.</param>
    /// <param name="message">Optional success message.</param>
    /// <typeparam name="T">The type of the data.</typeparam>
    /// <returns>A successful ApiResult instance containing the data.</returns>
    protected ApiResult<T> SuccessResult<T>(T data, string? message = default)
    {
        return new ApiResult<T>
        {
            Success = true,
            Code = ApiCode.Success,
            Message = message,
            Data = data,
        };
    }

    /// <summary>
    /// Creates a successful paged API result with data.
    /// </summary>
    /// <param name="data">The data to include in the result.</param>
    /// <param name="pageNumber">The current page number.</param>
    /// <param name="pageSize">The number of items per page.</param>
    /// <param name="totalCount">The total number of items.</param>
    /// <typeparam name="T">The type of the data.</typeparam>
    /// <returns>A successful ApiResult instance containing the paged data.</returns>
    protected ApiResult<PagedResult<T>> SuccessPagedResult<T>(
        IEnumerable<T> data,
        int pageNumber,
        int pageSize,
        long totalCount)
    {
        var totalPages = (int)Math.Ceiling(totalCount / (double)pageSize);

        var pagedResult = new PagedResult<T>
        {
            Data = data,
            PageNumber = pageNumber,
            PageSize = pageSize,
            TotalCount = totalCount,
            TotalPages = totalPages,
        };

        return new ApiResult<PagedResult<T>>
        {
            Success = true,
            Code = ApiCode.Success,
            Data = pagedResult,
        };
    }

    /// <summary>
    /// Creates a bad request API result.
    /// </summary>
    /// <param name="message">Optional error message.</param>
    /// <returns>A bad request ApiResult instance.</returns>
    protected ApiResult BadRequestResult(string? message = default)
    {
        return new ApiResult
        {
            Success = false,
            Code = ApiCode.InvalidRequest,
            Message = message,
        };
    }

    /// <summary>
    /// Creates a bad request API result with data.
    /// </summary>
    /// <param name="data">The data to include in the result.</param>
    /// <param name="message">Optional error message.</param>
    /// <typeparam name="T">The type of the data.</typeparam>
    /// <returns>A bad request ApiResult instance containing the data.</returns>
    protected ApiResult<T> BadRequestResult<T>(T data, string? message = default)
    {
        return new ApiResult<T>
        {
            Success = false,
            Code = ApiCode.InvalidRequest,
            Message = message,
            Data = data,
        };
    }

    /// <summary>
    /// Creates a failure API result.
    /// </summary>
    /// <param name="code">The error code.</param>
    /// <param name="message">Optional error message.</param>
    /// <returns>A failure ApiResult instance.</returns>
    protected ApiResult FailureResult(ApiCode code, string? message = default)
    {
        return new ApiResult
        {
            Success = false,
            Code = code,
            Message = message,
        };
    }

    /// <summary>
    /// Creates a failure API result with data.
    /// </summary>
    /// <param name="code">The error code.</param>
    /// <param name="data">The data to include in the result.</param>
    /// <param name="message">Optional error message.</param>
    /// <typeparam name="T">The type of the data.</typeparam>
    /// <returns>A failure ApiResult instance containing the data.</returns>
    protected ApiResult<T> FailureResult<T>(ApiCode code, T data, string? message = default)
    {
        return new ApiResult<T>
        {
            Success = false,
            Code = code,
            Message = message,
            Data = data,
        };
    }

    /// <summary>
    /// Creates a no data found API result.
    /// </summary>
    /// <param name="message">Optional message indicating no data was found.</param>
    /// <returns>A no data found ApiResult instance.</returns>
    protected ApiResult NoDataFoundResult(string? message = default)
    {
        return new ApiResult
        {
            Success = true,
            Code = ApiCode.NoDataFound,
            Message = message ?? "No data found.",
        };
    }

    /// <summary>
    /// Creates a no data found API result with data.
    /// </summary>
    /// <param name="message">Optional message indicating no data was found.</param>
    /// <typeparam name="T">The type of the data.</typeparam>
    /// <returns>A no data found ApiResult instance containing the data.</returns>
    protected ApiResult<T> NoDataFoundResult<T>(string? message = default)
    {
        return new ApiResult<T>
        {
            Success = true,
            Code = ApiCode.NoDataFound,
            Message = message ?? "No data found.",
            Data = default,
        };
    }

    /// <summary>
    /// Creates a no data found API result with paged data.
    /// </summary>
    /// <param name="message">Optional message indicating no data was found.</param>
    /// <typeparam name="T">The type of the data.</typeparam>
    /// <returns>A no data found ApiResult instance containing the paged data.</returns>
    protected ApiResult<PagedResult<T>> NoDataFoundPagedResult<T>(string? message = default)
    {
        return new ApiResult<PagedResult<T>>
        {
            Success = true,
            Code = ApiCode.NoDataFound,
            Message = message ?? "No data found.",
            Data = new PagedResult<T>
            {
                Data = [],
                PageNumber = 1,
                PageSize = 0,
                TotalCount = 0,
                TotalPages = 0,
            },
        };
    }

    /// <summary>
    /// Creates an error API result.
    /// </summary>
    /// <param name="message">Optional error message.</param>
    /// <returns>An error ApiResult instance.</returns>
    protected ApiResult ErrorResult(string? message = default)
    {
        return new ApiResult
        {
            Success = false,
            Code = ApiCode.UnknownError,
            Message = message,
        };
    }

    /// <summary>
    /// Creates an error API result with data.
    /// </summary>
    /// <param name="message">Optional error message.</param>
    /// <typeparam name="T">The type of the data.</typeparam>
    /// <returns>An error ApiResult instance containing the data.</returns>
    protected ApiResult<T> ErrorResult<T>(string? message = default)
    {
        return new ApiResult<T>
        {
            Success = false,
            Code = ApiCode.UnknownError,
            Message = message,
            Data = default,
        };
    }

    /// <summary>
    /// Creates an error API result with data.
    /// </summary>
    /// <param name="data">The data to include in the result.</param>
    /// <param name="message">Optional error message.</param>
    /// <typeparam name="T">The type of the data.</typeparam>
    /// <returns>An error ApiResult instance containing the data.</returns>
    protected ApiResult<T> ErrorResult<T>(T data, string? message = default)
    {
        return new ApiResult<T>
        {
            Success = false,
            Code = ApiCode.UnknownError,
            Message = message,
            Data = data,
        };
    }
}

File: src/nodejs/src/services/BaseService.ts
import { ApiCode } from "../dtos/ApiCode";
import { ApiDataResult, ApiResult } from "../dtos/ApiResult";
import { PagedResult } from "../dtos/PagedResult";

/**
 * Base class for services providing common helper methods.
 */
export abstract class BaseService {
  /**
   * Creates a successful API result.
   * @param message Optional success message.
   * @returns A successful ApiResult instance.
   */
  protected successResult(message?: string): ApiResult {
    return new ApiResult(true, ApiCode.Success, message || "Success");
  }

  /**
   * Creates a successful API result with data.
   * @param data The data to include in the result.
   * @param message Optional success message.
   * @returns A successful ApiDataResult instance containing the data.
   */
  protected successDataResult<T>(data: T, message?: string): ApiDataResult<T> {
    return new ApiDataResult<T>(true, ApiCode.Success, message || "Success", data);
  }

  /**
   * Creates a successful paged API result with data.
   * @param data The data to include in the result.
   * @param totalCount The total number of items.
   * @param pageNumber The current page number.
   * @param pageSize The number of items per page.
   * @param message Optional success message.
   * @returns A successful ApiDataResult instance containing the paged data.
   */
  protected successPagedResult<T>(
    data: T[],
    totalCount: number,
    pageNumber: number,
    pageSize: number,
    message?: string
  ): ApiDataResult<PagedResult<T>> {
    const pagedResult = new PagedResult(data, totalCount, pageNumber, pageSize);
    return new ApiDataResult<PagedResult<T>>(true, ApiCode.Success, message || "Success", pagedResult);
  }

  /**
   * Creates a bad request API result.
   * @param message Optional error message.
   * @returns A bad request ApiResult instance.
   */
  protected badRequestResult(message?: string): ApiResult {
    return new ApiResult(false, ApiCode.InvalidRequest, message || "Bad request");
  }

  /**
   * Creates a bad request API result with data.
   * @param message Optional error message.
   * @returns A bad request ApiDataResult instance.
   */
  protected badRequestDefaultDataResult<T>(message?: string): ApiDataResult<T> {
    return new ApiDataResult<T>(false, ApiCode.InvalidRequest, message || "Bad request", undefined);
  }

  /**
   * Creates a bad request API result with data.
   * @param data Optional data to include in the result.
   * @param message Optional error message.
   * @returns A bad request ApiDataResult instance.
   */
  protected badRequestDataResult<T>(data?: T, message?: string): ApiDataResult<T> {
    return new ApiDataResult<T>(false, ApiCode.InvalidRequest, message || "Bad request", data);
  }

  /**
   * Creates a failure API result.
   * @param code The error code.
   * @param message Optional error message.
   * @returns A failure ApiResult instance.
   */
  protected failureResult(code: ApiCode, message?: string): ApiResult {
    return new ApiResult(false, code, message || "Operation failed");
  }

  /**
   * Creates a failure API result with data.
   * @param code The error code.
   * @param data Optional data to include in the result.
   * @param message Optional error message.
   * @returns A failure ApiDataResult instance.
   */
  protected failureDataResult<T>(code: ApiCode, data?: T, message?: string): ApiDataResult<T> {
    return new ApiDataResult<T>(false, code, message || "Operation failed", data);
  }

  /**
   * Creates a no data found API result.
   * @param message Optional message indicating no data was found.
   * @returns A no data found ApiResult instance.
   */
  protected noDataFoundResult(message?: string): ApiResult {
    return new ApiResult(true, ApiCode.NoDataFound, message || "No data found");
  }

  /**
   * Creates a no data found API result with data.
   * @param data Optional data to include in the result.
   * @param message Optional message indicating no data was found.
   * @returns A no data found ApiDataResult instance.
   */
  protected noDataFoundDataResult<T>(data?: T, message?: string): ApiDataResult<T> {
    return new ApiDataResult<T>(true, ApiCode.NoDataFound, message || "No data found", data);
  }

  /**
   * Creates a no data found paged API result.
   * @param message Optional message indicating no data was found.
   * @returns A no data found ApiDataResult instance with empty paged result.
   */
  protected noDataFoundPagedResult<T>(message?: string): ApiDataResult<PagedResult<T>> {
    const emptyPagedResult = new PagedResult<T>([], 0, 1, 0);
    return new ApiDataResult<PagedResult<T>>(true, ApiCode.NoDataFound, message || "No data found", emptyPagedResult);
  }

  /**
   * Creates an error API result.
   * @param message Optional error message.
   * @returns An error ApiResult instance.
   */
  protected errorResult(message?: string): ApiResult {
    return new ApiResult(false, ApiCode.UnknownError, message || "An error occurred");
  }

  /**
   * Creates an error API result with data.
   * @param data Optional data to include in the result.
   * @param message Optional error message.
   * @returns An error ApiDataResult instance.
   */
  protected errorDataResult<T>(data?: T, message?: string): ApiDataResult<T> {
    return new ApiDataResult<T>(false, ApiCode.UnknownError, message || "An error occurred", data);
  }
}
